config/mtls.go

```diff
@@ -192,6 +192,30 @@ func (s *DownstreamMTLSSettings) applySettingsProto(
 	s.Enforcement = mtlsEnforcementFromProtoEnum(ctx, p.Enforcement)
 }
 
+func (s *DownstreamMTLSSettings) toSettingsProto() *config.DownstreamMtlsSettings {
+	var settings config.DownstreamMtlsSettings
+	if s.CA != "" || s.CAFile != "" {
+		settings.Ca = valueOrFromFileBase64(s.CA, s.CAFile)
+	}
+	if s.CRL != "" || s.CRLFile != "" {
+		settings.Crl = valueOrFromFileBase64(s.CRL, s.CRLFile)
+	}
+	if s.Enforcement != "" {
```

Comment:
Agreed

Reply:
After considering this some more, I think it will be best not to set default values here, because we want the ability to convert back and obtain an identical message to the original input. By setting defaults, we would be discarding information about the original message that could potentially alter its semantics (for example in this case, because the default value only "means something" if there is a valid CA, otherwise it is ignored - but that is an implementation detail)

Reply:
I think this is a tricky question in general. One guiding question — if we ever change a default setting in Core, what would need to happen in the Zero import flow?

I'd like for Zero to be as oblivious as possible to the Core version — I wouldn't want to start building logic into the import method like "if the incoming request is Core <= v0.27, then make sure to treat the default enforcement mode as XYZ, or else ...."

For this reason I would expect the config proto to represent "what is the current effective configuration" rather than "what was initially in the config.yaml file."

In this specific case, I would expect any no-op mTLS settings, like
```yaml
downstream_mtls:
  enforcement_mode: reject_connection
```

or

```yaml
downstream_mtls: {}
```

to translate to an unset `downstream_mtls` field in the config.proto representation.

Reply:
Right, in that case we want to leave those fields empty/unset, to let different versions of core decide how they interpret an unset field (by picking a default). If we set them at the time of import, then change the default value, it won't end up changing because we would have stored the default value at the time of import.

Reply:
OK, then I think we disagree on what the intended behavior is. My expectation is that once a configuration is managed by Pomerium Zero, it should not be changed except by the admin user, whether that's through the Pomerium Zero console or API. I would not expect the Core defaults to figure into it at all at that point.

Reply:
I'm not sure I understand what you mean, for example if you have 
```yaml
downstream_mtls:
  ca_file: ca.crt
```
should this translate into 
```
downstream_mtls: {
  ca: <...>
  enforcement: POLICY_WITH_DEFAULT_DENY
}
```
or
```
downstream_mtls: {
  ca: <...>
}
```

Reply:
I would expect the former:
```
downstream_mtls: {
  ca: <...>
  enforcement: POLICY_WITH_DEFAULT_DENY
}
```

(i.e. the configuration is specified explicitly)

But maybe there are other cases where this doesn't make sense? Please let me if any come to mind — it's entirely possible that I'm missing something.

Reply:
Runtime flags is one example where we wouldn't want to do this. omitting the `runtime_flags` field _should_ have different behavior based on the version of pomerium you're running, since we can add/remove these like feature gates.

In general, core is allowed to construct default values based on any information it wants, including other fields, environment specifics, etc. The default values are not necessarily static, and omitting a field can signal a different intent than setting it to what the default value would be. I think we would want to preserve this logic as much as possible to avoid unexpected behavior.

Reply:
1. Thanks for bringing up runtime flags; that's an interesting example. I think so far we haven't allowed these to be set from Pomerium Zero. How do we want to handle those during import? Should we emit a warning/error if any are set to a non-default value?

2.  I think we still disagree on what would be expected or unexpected.

    > core is allowed to construct default values based on any information it wants, including other fields, environment specifics, etc.

    I don't think I understand this. In my mental model, all the default settings _are_ static, and are documented (or at least _should_ be documented) on the corresponding reference pages on the docs site.

    Do you have any examples in mind where this isn't the case?

Reply:
> How do we want to handle those during import? Should we emit a warning/error if any are set to a non-default value?
We will need to either support them in zero or disallow import entirely if any are present.

Reply:
>Do you have any examples in mind where this isn't the case?

There is so much logic around loading and handling of default options that I can't be sure _any_ of them are truly static.